add foreign key doc (#12275)
pingcap/docs: sql-statements/sql-statement-foreign-key.md

---
title: FOREIGN KEY
summary: An overview of the usage of FOREIGN KEY for the TiDB database.
---

# FOREIGN KEY

Starting from v6.6.0, TiDB supports the foreign key feature, which allows cross-table referencing of related data, and foreign key constraints to maintain data consistency.

The foreign key is defined in the child table. The syntax is as follows:

> ```sql

Reviewer: I think we should use `ebnf+diagram` here?

The effect of disabling [`foreign_key_checks`](/system-variables.md#foreign_key_checks) is as follows:

- When deleting a parent table referenced by a foreign key, the deletion can succeed only when the foreign key constraint check is disabled.
- When importing data to a database, the order of creating tables might be different from the foreign key dependency order, which might cause the creation of tables to fail. Only when the foreign key constraint check is disabled can the tables be created successfully. In addition, when importing data, disabling the foreign key constraint check can also speed up data import.
- When importing data to a database, if the data of the child table is imported first, an error will be reported. Only when the foreign key constraint check is disabled can the data of the child table be imported successfully.

> - When executing `ALTER TABLE` operations related to foreign keys, the foreign key constraint check must be enabled to execute successfully.

Reviewer 2: ```suggestion
- If an `ALTER TABLE` operation to be executed involves change of the foreign key, this operation succeeds only when the foreign key constraint check is disabled.
```
It seems "关于" in Chinese should be "关闭".

When the foreign key constraint check is disabled, the foreign key constraint check and reference operation are not executed, except for the following scenarios:

> - When executing `ALTER TABLE`, if the definition of the foreign key is incorrect, an error is still reported.

Reviewer 2: Inconsistent with the Chinese. Please confirm. If we follow the Chinese, the translation should be:
> If the execution of  `ALTER TABLE` might result in wrong definition of the foreign key, an error is still reported during the execution.

Reviewer 3: I already change the Chinese version, and this is lgtm.

> - When deleting the index required by the foreign key, the foreign key must be deleted first. Otherwise, an error is reported when deleting the foreign key.

Reviewer 2: ```suggestion
- When deleting the index required by the foreign key, you should delete the foreign key first. Otherwise, an error is reported.
```

> - When creating a foreign key, if it does not meet the conditions or restrictions of the foreign key, an error is reported.

Reviewer 2: ```suggestion
- When you create a foreign key that does not meet related conditions or restrictions, an error is reported.
```

> ## Lock

Reviewer: ```suggestion
## Locking
```

When `INSERT` or `UPDATE` a child table, the foreign key constraint checks whether the corresponding foreign key value exists in the parent table, and locks the row in the parent table to avoid the foreign key value being deleted by other operations violating the foreign key constraint. The locking behavior is equivalent to performing a `SELECT FOR UPDATE` operation on the row where the foreign key value is located in the parent table. Because TiDB currently does not support `LOCK IN SHARE MODE`, if there are a lot of concurrent writes to the child table and most of the referenced foreign key values are the same, there might be serious lock conflicts. It is recommended to disable [`foreign_key_checks`](/system-variables.md#foreign_key_checks) when writing a large number of child table data.

> ## Definition and metadata of foreign keys

Reviewer: Looks like `information_schema.table_constraints` should also be mentioned.

```
sql> SELECT * FROM information_schema.table_constraints WHERE CONSTRAINT_TYPE='FOREIGN KEY'\G
*************************** 1. row ***************************
CONSTRAINT_CATALOG: def
 CONSTRAINT_SCHEMA: test
   CONSTRAINT_NAME: fk_1
      TABLE_SCHEMA: test
        TABLE_NAME: t2
   CONSTRAINT_TYPE: FOREIGN KEY
1 row in set (0.0025 sec)
```

And `information_schema.referential_constraints`:
```
sql> TABLE information_schema.referential_constraints\G
*************************** 1. row ***************************
       CONSTRAINT_CATALOG: def
        CONSTRAINT_SCHEMA: test
          CONSTRAINT_NAME: fk_1
UNIQUE_CONSTRAINT_CATALOG: def
 UNIQUE_CONSTRAINT_SCHEMA: test
   UNIQUE_CONSTRAINT_NAME: PRIMARY
             MATCH_OPTION: NONE
              UPDATE_RULE: NO ACTION
              DELETE_RULE: CASCADE
               TABLE_NAME: t2
    REFERENCED_TABLE_NAME: t1
1 row in set (0.0021 sec)
```

Reviewer 3: Great, Nice catch, I will add those in https://github.com/pingcap/docs-cn/pull/11974 first.

To view the definition of a foreign key constraint, use the `SHOW CREATE TABLE` statement:

```sql
mysql> SHOW CREATE TABLE child\G
*************************** 1. row ***************************
       Table: child
Create Table: CREATE TABLE `child` (
  `id` int(11) DEFAULT NULL,
  `pid` int(11) DEFAULT NULL,
  KEY `idx_pid` (`pid`),
  CONSTRAINT `fk_1` FOREIGN KEY (`pid`) REFERENCES `test`.`parent` (`id`) ON DELETE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
```

> You can get information about foreign keys from the `INFORMATION_SCHEMA.KEY_COLUMN_USAGE` system table. The following is an example query:

Reviewer: ```suggestion
You can get information about foreign keys from the [`INFORMATION_SCHEMA.KEY_COLUMN_USAGE`](information-schema/information-schema-key-column-usage.md) system table. The following is an example query:
```

```sql
mysql> SELECT TABLE_SCHEMA, TABLE_NAME, COLUMN_NAME, CONSTRAINT_NAME FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE WHERE REFERENCED_TABLE_SCHEMA IS NOT NULL;
+--------------+---------------+------------------+-----------------+
| TABLE_SCHEMA | TABLE_NAME    | COLUMN_NAME      | CONSTRAINT_NAME |
+--------------+---------------+------------------+-----------------+
| test         | child         | pid              | fk_1            |
| test         | product_order | product_category | fk_1            |
| test         | product_order | product_id       | fk_1            |
| test         | product_order | customer_id      | fk_2            |
+--------------+---------------+------------------+-----------------+
```

> ## View the execution plan with foreign keys
> 
> You can use the `EXPLAIN` statement to view the execution plan. The `Foreign_Key_Check` operator is the operator that performs the foreign key constraint check when executing DML statements.

Reviewer 2: ```suggestion
## View execution plans with foreign keys

You can use the `EXPLAIN` statement to view execution plans. The `Foreign_Key_Check` operator performs the foreign key constraint check on DML statements that are executed.
```

```sql
mysql> explain insert into child values (1,1);
+-----------------------+---------+------+---------------+-------------------------------+
| id                    | estRows | task | access object | operator info                 |
+-----------------------+---------+------+---------------+-------------------------------+
| Insert_1              | N/A     | root |               | N/A                           |
| └─Foreign_Key_Check_3 | 0.00    | root | table:parent  | foreign_key:fk_1, check_exist |
+-----------------------+---------+------+---------------+-------------------------------+
```

> You can use the `EXPLAIN ANALYZE` statement to view the execution of the foreign key reference behavior. The `Foreign_Key_Cascade` operator is the operator that performs the foreign key reference behavior when executing DML statements.

Reviewer 2: ```suggestion
You can use the `EXPLAIN ANALYZE` statement to view the execution of the foreign key reference behavior. The `Foreign_Key_Cascade` operator performs foreign key referencing for DML statements that are executed.
```

|       ├─IndexRangeScan_9(Build)  | 10.00   | 1       | cop[tikv] | table:child, index:idx_pid(pid) | time:351.2µs, loops:3, cop_task: {num: 1, max: 282.3µs, proc_keys: 0, rpc_num: 1, rpc_time: 263µs, copr_cache_hit_ratio: 0.00, distsql_concurrency: 15}, tikv_task:{time:220.2µs, loops:0}   | range:[1,1], keep order:false, stats:pseudo | N/A       | N/A  |
|       └─TableRowIDScan_10(Probe) | 10.00   | 1       | cop[tikv] | table:child                     | time:223.9µs, loops:2, cop_task: {num: 1, max: 168.8µs, proc_keys: 0, rpc_num: 1, rpc_time: 154.5µs, copr_cache_hit_ratio: 0.00, distsql_concurrency: 15}, tikv_task:{time:145.6µs, loops:0} | keep order:false, stats:pseudo              | N/A       | N/A  |
+----------------------------------+---------+---------+-----------+---------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------+-----------+------+
```

## Compatibility

### Compatibility between TiDB versions

> Before v6.6.0, TiDB supported the syntax of creating foreign keys, but the created foreign keys are ineffective. If you upgrade a TiDB cluster created before v6.6.0 to v6.6.0 or later, the foreign keys created before the upgrade are still ineffective. Only the foreign keys created in v6.6.0 or later versions are effective. You can use the `SHOW CREATE TABLE` statement to check whether the foreign keys are effective. The invalid foreign key has a `/* FOREIGN KEY INVALID */` comment.

Reviewer 2: ```suggestion
Before v6.6.0, TiDB supports the syntax of creating foreign keys, but the created foreign keys are ineffective. If you upgrade a TiDB cluster created before v6.6.0 to v6.6.0 or later, the foreign keys created before the upgrade are still ineffective. Only the foreign keys created in v6.6.0 or later versions are effective. You can use the `SHOW CREATE TABLE` statement to check whether the foreign keys are effective. An invalid foreign key has a `/* FOREIGN KEY INVALID */` comment.
```

```sql
mysql> SHOW CREATE TABLE child\G
***************************[ 1. row ]***************************
Table        | child
Create Table | CREATE TABLE `child` (
  `id` int(11) DEFAULT NULL,
  `pid` int(11) DEFAULT NULL,
  KEY `idx_pid` (`pid`),
  CONSTRAINT `fk_1` FOREIGN KEY (`pid`) REFERENCES `test`.`parent` (`id`) ON DELETE CASCADE /* FOREIGN KEY INVALID */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
```

### Compatibility with MySQL

> - [TiDB Binlog](/tidb-binlog/tidb-binlog-overview.md) does not support foreign keys.

Reviewer: This doesn't below under a "Compatibility with MySQL" header. Maybe put this on a different header or rename the section

Reviewer: Maybe also document compatibility with Dumpling, Lightning, DM, CDC and sync-diff-inspector etc.

```

## Compatibility

### Compatibility between TiDB versions

Before v6.6.0, TiDB supports the syntax of creating foreign keys, but the created foreign keys are ineffective. If you upgrade a TiDB cluster created before v6.6.0 to v6.6.0 or later, the foreign keys created before the upgrade are still ineffective. Only the foreign keys created in v6.6.0 or later versions are effective. You can use the `SHOW CREATE TABLE` statement to check whether the foreign keys are effective. The invalid foreign key has a `/* FOREIGN KEY INVALID */` comment.

```sql
mysql> SHOW CREATE TABLE child\G
***************************[ 1. row ]***************************
Table        | child
Create Table | CREATE TABLE `child` (
  `id` int(11) DEFAULT NULL,
  `pid` int(11) DEFAULT NULL,
  KEY `idx_pid` (`pid`),
  CONSTRAINT `fk_1` FOREIGN KEY (`pid`) REFERENCES `test`.`parent` (`id`) ON DELETE CASCADE /* FOREIGN KEY INVALID */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
```

> ### Compatibility with TiDB tools

Author: @qiancai PTAL
